app/src/org/commcare/appupdate/CommcareFlexibleAppUpdateManager.java

```diff
@@ -0,0 +1,206 @@
+        HiddenPreferences.skipCommCareVersionForUpdate(currentAvailableVersion);
+    }
+
+    //endregion
+
+    @Override
+    public void onStateUpdate(InstallState state) {
+        Logger.log(TAG, "Install state updated with install status: " + state.installStatus()
+                + ", and error code: " + state.installErrorCode());
+        mInstallErrorCode = state.installErrorCode();
+        if (state.installStatus() == InstallStatus.DOWNLOADING) {
+            // TODO: Should be keep on publishing status for showing download percentage.
+            // From https://stackoverflow.com/a/10415638/6671572
+            percentageDownloaded = (int) (state.bytesDownloaded() * 100.0 / state.totalBytesToDownload() + 0.5) ;
+        }
+        if (mInstallStatus != state.installStatus()) {
+            mInstallStatus = state.installStatus();
+            publishStatus();
+        }
+    }
+
+    //region Private helpers
+    private void fetchAppUpdateInfo() {
+        mAppUpdateManager.getAppUpdateInfo()
+                .addOnSuccessListener(info -> {
+                    mAppUpdateInfo = info;
+                    mUpdateAvailability = info.updateAvailability();
+                    mInstallStatus = info.installStatus();
+                    Logger.log(TAG, "fetchAppUpdateInfo|success with UpdateAvailability : "
+                            + mUpdateAvailability + ", and InstallStatus : " + mInstallStatus);
+                    publishStatus();
+                })
+                .addOnFailureListener(exception -> {
+                    mAppUpdateInfo = null;
+                    mUpdateAvailability = UpdateAvailability.UNKNOWN;
+                    mInstallStatus = InstallStatus.UNKNOWN;
+                    Logger.log(TAG, "fetchAppUpdateInfo|failed with : " + exception.getMessage());
+                    publishStatus();
+                });
+    }
+
+    /**
+     * Pings the listener using {@link #mCallback} when there is an update regarding installation state.
+     */
+    private void publishStatus() {
+        if (!mEnabled) {
+            return;
+        }
+        AppUpdateState newState = getAppUpdateState();
+        if (mAppUpdateState == newState) {
+            return;
+        }
+        Logger.log(TAG, "Publishing status update to : " + newState.name() + ", from : "
+                + (mAppUpdateState != null ? mAppUpdateState.name() : "null"));
+        mAppUpdateState = newState;
+        mCallback.run();
+    }
+
+    /**
+     * Gets {@link AppUpdateState} using {@link #mUpdateAvailability} and {@link #mInstallStatus}
+     */
+    private AppUpdateState getAppUpdateState() {
+        if (mAppUpdateInfo != null && mAppUpdateInfo.availableVersionCode() == HiddenPreferences.getSkippedCommCareUpdateVersion()) {
+            return AppUpdateState.UNAVAILABLE;
+        }
+        AppUpdateState state = AppUpdateState.UNAVAILABLE;
+        switch (mInstallStatus) {
+            case InstallStatus.PENDING:
```

Comment:
Confirming `Pending` doesn't mean that an app update is available but not downloading ?

Reply:
Ohh, I missed it. The `Pending` status means that the downloading is paused maybe due to internet issues.